### ISSUE TITLE
Recomputing automatic derivatives in pyscfad

### ISSUE THREAD

**Issue by deyanmihaylov** [fishjojo/pyscfad#13]
Together with Nike Dattani we have been using the automatic derivatives functionality of `pyscfad`. I have a question, which I have been thinking about for some days.

When computing the energy for a given atom, e.g. Ne:
```
mol = gto.Mole()
mol.atom = 'Ne 0 0 0'  # in Angstrom

mol.basis = {'Ne': pyscf.gto.basis.parse("""
Ne S
      2.5 1
      
Ne S
     0.4 1""")}

mol.verbose = VERBOSITY
mol.build()

mf = scf.RHF(mol)
e = mf.kernel()
```

If then I would like to compute the derivative of the energy wrt the exponents, I'd need to do:
```
jac = mf.energy_grad()
grad_E = np.array(jac.exp)
```
and the derivatives are in `grad_E`. However, say that I would like to compute the derivatives wrt different values of the exponents. Is there a way of doing it WITHOUT building the atom from scratch with different basis? I suppose that the most expensive part of the code is actually computing the automatic derivatives. Instead, it would be great if we could just evaluate the already computed autodiffs at different values of the exponents, given that changing just the values of the exponents does not change the functional form of the energy (at least as much as I understand what's going on in the package). Is that currently possible?

**Comment by ndattani** [comment]
If I understand you correctly, you are saying that if the Hartree-Fock energy "E" as a function of an exponent "x" is E(x), then to get dE(1.34)/dx and dE(2.67)/dx, you are calculating dE(x)/dx twice:
```
deriv_val1 = jax.jacrev(E)(1.34)
deriv_val2 = jax.jacrev(E)(2.67)
```
But instead you want to calculate dE(x)/dx once, then store it as F(x), and just evaluate F(1.34) and F(2.67):
```
deriv_f = jax.jacrev(E)
deriv_val1 = deriv_E(1.34)
deriv_val2 = deriv_E(2.67)
```
If you are trying to optimize an exponent, it would be cheaper to just calculate the functional form of dE(x)/dX once, then evaluate it 100s of times whenever the optimizer changes the guess value for the exponent, rather than recalculating the auto-derivatives hundreds of times.

Is that right @deyanmihaylov?

**Comment by deyanmihaylov** [comment]
That is what I am saying @ndattani thank you for the comment.

**Comment by fishjojo** [comment]
This should already be taken care of by Jax. If you try the following example, you will find that the first gradient calculation is slower than the rest. For the first gradient calculation, Jax will first perform 'autodiff' abstractly, i.e. tracing the computation graph and cache/optimize the corresponding functions. And for the rest of gradient calculations, the cached functions are used. But to get the actual gradient, one still needs to go through the forward pass, which computes the necessary intermediate quantities, and then backpropagate to compute the gradient. It's the same as analytic evaluation of gradients where one still needs to recompute the SCF energy whenever the basis is changed. 
```
import jax
import pyscf
from pyscf.lib import logger
from pyscfad import gto, scf

bases = [
{'Ne': pyscf.gto.basis.parse("""
    Ne S
          2.5 1

    Ne S
         0.4 1""")},
{'Ne': pyscf.gto.basis.parse("""
    Ne S
          2.6 1

    Ne S
         0.4 1""")},
{'Ne': pyscf.gto.basis.parse("""
    Ne S
          2.7 1

    Ne S
         0.4 1""")},
]

def build_mol(basis):
    mol = gto.Mole()
    mol.atom = 'Ne 0 0 0'
    mol.basis = basis
    mol.verbose = 5
    mol.build(trace_coords=False, trace_exp=True, trace_ctr_coeff=False)
    return mol

mols = []
for basis in bases:
    mols.append(build_mol(basis))

def energy(mol):
    mf = scf.RHF(mol)
    e = mf.kernel()
    return e

jacs = []
for i, mol in enumerate(mols):
    cput0 = (logger.process_clock(), logger.perf_counter())
    jacs.append(jax.grad(energy)(mol))
    logger.timer(mol, f'gradient for mol {i+1}', *cput0)

for jac in jacs:
    print(jac.exp)
```

**Comment by ndattani** [comment]
@deyanmihaylov can you let us know if this answers your question? Or is your problem more complicated (e.g. involving the Hessian)?
